### PR TITLE
Move collective pipeliner after post-layout opts and add a unit test

### PR DESCRIPTION
**Pull request by wenscarl**
@kaixih @frgossen This is just a rebased version of https://github.com/openxla/xla/pull/12866

### CODE REVIEW

**Review by frgossen — APPROVED** [review]

File: xla/tests/collective_ops_test_e2e.cc [hunk 1]
```diff
@@ -855,5 +876,98 @@ ENTRY main.9_spmd {
...
+      auto module, ParseAndReturnVerifiedModule(kModuleReplicatedStr, config));
+
+  TF_ASSERT_OK_AND_ASSIGN(auto executable,
+                          CreateExecutable(std::move(module),
+                                           /*run_hlo_passes=*/true));
+  EXPECT_TRUE(executable->has_module());
+  HloInstruction* gemm_op =
+      FindInstruction(&executable->module(), HloOpcode::kCustomCall);
+  EXPECT_THAT(gemm_op, NotNull());
+  EXPECT_TRUE(gemm_op->custom_call_target() == "__cublas$lt$matmul$f8");
```
Comment (akuegel) [inline xla/tests/collective_ops_test_e2e.cc:970]: Can you please use EXPECT_EQ(gemm_op->custom_call_target(), "__cublas$lt$matmul$f8") ?
The linter complains about this line.

File: xla/tests/collective_ops_test_e2e.cc [hunk 2]
```diff
@@ -24,6 +24,7 @@ limitations under the License.
 #include "xla/hlo/utils/hlo_matchers.h"
```
Comment (akuegel) [inline xla/tests/collective_ops_test_e2e.cc:24]: Could you maybe rename this test file to collective_ops_e2e_test.cc? Our infra is setup to check for test.cc suffix for verifying that a change contains also a test change.

Comment (penpornk) [reply xla/tests/collective_ops_test_e2e.cc:24]: I have just changed that in https://github.com/openxla/xla/commit/812d0f931cd201ead20c86d717e5555ea2e96250. 
@wenscarl Could you please help rebase?

Comment (wenscarl) [reply xla/tests/collective_ops_test_e2e.cc]: Rebase is done.

**Review by penpornk — APPROVED** [review]
Thank you for merging conflicts!

**Comment by kaixih** [comment]
It seems the merging is blocked. By browsing the log, I don't think it is related to this PR. @penpornk Can you help with it?

**Pull request closed without merging**